server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -0,0 +1,397 @@
+            segmentsAndMetadata.getCommitMetadata(),
+            Joiner.on(", ").join(segmentsAndMetadata.getSegments())
+        );
+
+        final boolean published = publisher.publishSegments(
+            ImmutableSet.copyOf(segmentsAndMetadata.getSegments()),
+            segmentsAndMetadata.getCommitMetadata()
+        );
+
+        if (published) {
+          log.info("Published segments, awaiting handoff.");
+        } else {
+          log.info("Transaction failure while publishing segments, checking if someone else beat us to it.");
+          if (usedSegmentChecker.findUsedSegments(segmentsToIdentifiers(segmentsAndMetadata.getSegments()))
+                                .equals(Sets.newHashSet(segmentsAndMetadata.getSegments()))) {
+            log.info("Our segments really do exist, awaiting handoff.");
+          } else {
+            log.warn("Our segments don't exist, giving up.");
+            return null;
+          }
+        }
+
+        for (final DataSegment dataSegment : segmentsAndMetadata.getSegments()) {
+          handoffNotifier.registerSegmentHandoffCallback(
+              new SegmentDescriptor(
+                  dataSegment.getInterval(),
+                  dataSegment.getVersion(),
+                  dataSegment.getShardSpec().getPartitionNum()
+              ),
+              MoreExecutors.sameThreadExecutor(),
+              new Runnable()
+              {
+                @Override
+                public void run()
+                {
+                  final SegmentIdentifier identifier = SegmentIdentifier.fromDataSegment(dataSegment);
+                  log.info("Segment[%s] successfully handed off, dropping.", identifier);
+                  final ListenableFuture<?> dropFuture = appenderator.drop(identifier);
+                  Futures.addCallback(
+                      dropFuture,
+                      new FutureCallback<Object>()
+                      {
+                        @Override
+                        public void onSuccess(Object result)
+                        {
+                          synchronized (handoffMonitor) {
+                            handoffMonitor.notifyAll();
+                          }
+                        }
+
+                        @Override
+                        public void onFailure(Throwable e)
+                        {
+                          log.warn(e, "Failed to drop segment[%s]?!");
+                          synchronized (handoffMonitor) {
+                            handoffMonitor.notifyAll();
+                          }
+                        }
+                      }
+                  );
+                }
+              }
+          );
+        }
+
+        return segmentsAndMetadata;
+      }
+      catch (Exception e) {
+        final long sleepMillis = computeNextRetrySleep(++nTry);
+        log.warn(e, "Failed publishAll (try %d), retrying in %,dms.", nTry, sleepMillis);
+        Thread.sleep(sleepMillis);
+      }
```

Comment:
an interrupt will cause `publishAll` to throw an InterruptedException (due to the Thread.sleep), is that enough?